src/lib/arch/riscv/Instruction_execute.cc

```diff
@@ -0,0 +1,665 @@
+#include <cmath>
+#include <tuple>
+
+#include "InstructionMetadata.hh"
+#include "simeng/arch/riscv/Instruction.hh"
+
+namespace simeng {
+namespace arch {
+namespace riscv {
+
+/** Multiply `a` and `b`, and return the high 64 bits of the result.
+ * https://stackoverflow.com/a/28904636 */
+uint64_t mulhi(uint64_t a, uint64_t b) {
+  uint64_t a_lo = (uint32_t)a;
+  uint64_t a_hi = a >> 32;
+  uint64_t b_lo = (uint32_t)b;
+  uint64_t b_hi = b >> 32;
+
+  uint64_t a_x_b_hi = a_hi * b_hi;
+  uint64_t a_x_b_mid = a_hi * b_lo;
+  uint64_t b_x_a_mid = b_hi * a_lo;
+  uint64_t a_x_b_lo = a_lo * b_lo;
+
+  uint64_t carry_bit = ((uint64_t)(uint32_t)a_x_b_mid +
+                        (uint64_t)(uint32_t)b_x_a_mid + (a_x_b_lo >> 32)) >>
+                       32;
+
+  uint64_t multhi =
+      a_x_b_hi + (a_x_b_mid >> 32) + (b_x_a_mid >> 32) + carry_bit;
+
+  return multhi;
+}
+
+/** Extend 'bits' by value in position 'msb' of 'bits' (1 indexed) */
+uint64_t bitExtend(uint64_t bits, uint64_t msb) {
+  int64_t leftShift = bits << (64 - msb);
+  int64_t rightShift = leftShift >> (64 - msb);
+  return rightShift;
+}
+
+uint64_t signExtendW(uint64_t bits) { return bitExtend(bits, 32); }
+
+uint64_t zeroExtend(uint64_t bits, uint64_t msb) {
```

Comment:
is msb supposed to be in the range [0, 63] or [1, 64] ? Because if it is the former, it can cause UB if msb is 0.

Reply:
This was resolved but not sure what the solution was?